import sys
import os
from PyQt5 import uic
from PyQt5.QtWidgets import (QApplication, QMainWindow, QMessageBox, 
                             QPushButton, QHBoxLayout, QWidget)
from PyQt5.QtCore import Qt, QTimer
from PyQt5.QtGui import QPixmap, QPalette, QBrush, QFont

# Importar la clase del sidebar
from sidebar import MainWindow  # Aseg√∫rate de que el archivo se llame sidebar.py

class LoginWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        
        # Cargar el dise√±o desde la carpeta ui
        ui_path = os.path.join("ui", "login.ui")
        uic.loadUi(ui_path, self)
        
        # Configurar fondo manualmente
        self.setup_background()
        self.setup_logo()
        
        # Aplicar estilos CSS b√°sicos
        self.setup_styles()
        
        # Configurar placeholders simples
        self.setup_placeholders()
        
        # Conectar eventos
        self.setup_events()
        
        # Centrar ventana
        self.center_window()
        
        print("‚úÖ Ventana cargada correctamente")
    
    def setup_background(self):
        """Configurar la imagen de fondo manualmente"""
        fondo_path = os.path.join("img", "fondologin.png")
        if os.path.exists(fondo_path):
            try:
                palette = QPalette()
                pixmap = QPixmap(fondo_path)
                palette.setBrush(QPalette.Window, QBrush(pixmap.scaled(
                    self.size(), Qt.IgnoreAspectRatio, Qt.SmoothTransformation)))
                self.setPalette(palette)
                print("‚úÖ Fondo cargado desde:", fondo_path)
            except Exception as e:
                print(f"‚ùå Error cargando fondo: {e}")
        else:
            print(f"‚ö†Ô∏è No se encontr√≥ el fondo en: {fondo_path}")

    def setup_logo(self):
        """Configurar la imagen del logo"""
        logo_path = os.path.join("img", "logo.png")
        if os.path.exists(logo_path):
            try:
                if hasattr(self, 'labelLogo'):
                    pixmap = QPixmap(logo_path)
                    scaled_pixmap = pixmap.scaled(100, 100, Qt.KeepAspectRatio, Qt.SmoothTransformation)
                    self.labelLogo.setPixmap(scaled_pixmap)
                    self.labelLogo.setAlignment(Qt.AlignCenter)
                    print("‚úÖ Logo cargado desde:", logo_path)
                else:
                    print("‚ö†Ô∏è No se encontr√≥ labelLogo en el dise√±o")
            except Exception as e:
                print(f"‚ùå Error cargando logo: {e}")
        else:
            print(f"‚ö†Ô∏è No se encontr√≥ el logo en: {logo_path}")
            if hasattr(self, 'labelLogo'):
                self.labelLogo.setText("üêÑ")
                self.labelLogo.setStyleSheet("font-size: 48px; color: white;")
                print("‚úÖ Logo de texto configurado")

    def center_window(self):
        """Centrar la ventana en la pantalla"""
        screen = QApplication.primaryScreen().geometry()
        size = self.geometry()
        self.move(
            (screen.width() - size.width()) // 2,
            (screen.height() - size.height()) // 2
        )
    
    def setup_styles(self):
        """Aplicar estilos CSS b√°sicos"""
        style = """
        QMainWindow {
            background: transparent;
        }
        
        QLabel {
            color: #FFFFFF;
            font-family: 'Segoe UI', Arial;
        }
        
        QLineEdit {
            background-color: rgba(255, 255, 255, 0.9);
            border: 1px solid rgba(255, 255, 255, 0.3);
            border-radius: 8px;
            padding: 12px 15px;
            color: #1D2128;
            font-size: 14px;
            font-family: 'Segoe UI', Arial;
        }
        
        QPushButton#btnLogin {
            background-color: #a09172;
            border: 2px solid #a09172;
            color: #FFFFFF;
            border-radius: 8px;
            padding: 12px 25px;
            font-weight: bold;
            font-size: 15px;
        }
        
        QLabel#labelError {
            color: #e74c3c;
            font-size: 12px;
            font-weight: bold;
        }
        """
        self.setStyleSheet(style)
    
    def setup_placeholders(self):
        """Configurar placeholders simples"""
        try:
            if hasattr(self, 'lineEditUsuario'):
                self.lineEditUsuario.setPlaceholderText("Usuario")
                print("‚úÖ Placeholder usuario configurado")
            
            if hasattr(self, 'lineEditPassword'):
                self.lineEditPassword.setPlaceholderText("Contrase√±a")
                self.lineEditPassword.setEchoMode(self.lineEditPassword.Password)
                print("‚úÖ Placeholder contrase√±a configurado")
                
        except Exception as e:
            print(f"‚ùå Error configurando placeholders: {e}")
    
    def setup_events(self):
        """Conectar eventos de manera segura"""
        if hasattr(self, 'btnLogin'):
            self.btnLogin.clicked.connect(self.validar_login)
            print("‚úÖ Bot√≥n login conectado")
        else:
            print("‚ùå No se encontr√≥ btnLogin")
        
        if hasattr(self, 'lineEditUsuario'):
            self.lineEditUsuario.returnPressed.connect(self.validar_login)
            self.lineEditUsuario.setFocus()
            print("‚úÖ Campo usuario conectado")
        
        if hasattr(self, 'lineEditPassword'):
            self.lineEditPassword.returnPressed.connect(self.validar_login)
            print("‚úÖ Campo contrase√±a conectado")
    
    def resizeEvent(self, event):
        """Redimensionar el fondo cuando cambia el tama√±o de la ventana"""
        super().resizeEvent(event)
        self.setup_background()
    
    def validar_login(self):
        """Validar credenciales"""
        try:
            usuario = ""
            contrasena = ""
            
            if hasattr(self, 'lineEditUsuario'):
                usuario = self.lineEditUsuario.text().strip()
            
            if hasattr(self, 'lineEditPassword'):
                contrasena = self.lineEditPassword.text()
            
            print(f"üîê Validando: usuario='{usuario}', contrase√±a='{contrasena}'")
            
            # Validaci√≥n simple
            if usuario == "admin" and contrasena == "1234":
                self.mostrar_mensaje("‚úì Acceso concedido", "√©xito")
                QTimer.singleShot(1000, self.abrir_menu_principal)
            else:
                self.mostrar_mensaje("‚úó Usuario o contrase√±a incorrectos", "error")
                if hasattr(self, 'lineEditPassword'):
                    self.lineEditPassword.clear()
                    self.lineEditPassword.setFocus()
                
        except Exception as e:
            print(f"‚ùå Error al validar: {e}")
            self.mostrar_mensaje("Error en la validaci√≥n", "error")
    
    def mostrar_mensaje(self, mensaje, tipo):
        """Mostrar mensaje de error/√©xito"""
        try:
            if hasattr(self, 'labelError'):
                self.labelError.setText(mensaje)
                if tipo == "error":
                    self.labelError.setStyleSheet("color: #e74c3c;")
                else:
                    self.labelError.setStyleSheet("color: #2ecc71;")
                print(f"üì¢ Mensaje: {mensaje}")
        except Exception as e:
            print(f"‚ùå Error mostrando mensaje: {e}")
    
    def abrir_menu_principal(self):
        """Abrir ventana principal (sidebar)"""
        try:
            # Ocultar el login
            self.hide()
            
            # Crear y mostrar el sidebar
            self.sidebar = MainWindow()
            
            # Conectar la se√±al de cierre del sidebar para cerrar toda la aplicaci√≥n
            self.sidebar.destroyed.connect(self.cerrar_aplicacion)
            
            # Mostrar el sidebar
            self.sidebar.show()
            
            # Opcional: Maximizar la ventana del sidebar
            self.sidebar.showMaximized()
            
            print("‚úÖ Sidebar abierto correctamente")
            
        except Exception as e:
            print(f"‚ùå Error al abrir el sidebar: {e}")
            QMessageBox.critical(self, "Error", f"No se pudo abrir el men√∫ principal: {e}")
    
    def cerrar_aplicacion(self):
        """Cerrar toda la aplicaci√≥n cuando se cierra el sidebar"""
        self.close()

def main():
    app = QApplication(sys.argv)
    
    # Configurar fuente global
    font = QFont("Segoe UI", 10)
    app.setFont(font)
    
    # Cargar estilo QSS si existe
    style_path = "style.qss"
    if os.path.exists(style_path):
        try:
            with open(style_path, "r") as style_file:
                style_str = style_file.read()
                app.setStyleSheet(style_str)
                print("‚úÖ Estilo QSS cargado correctamente")
        except Exception as e:
            print(f"‚ùå Error cargando estilo QSS: {e}")
    
    # Verificar que los archivos existen
    ui_path = os.path.join("ui", "login.ui")
    fondo_path = os.path.join("img", "fondologin.png")
    
    if not os.path.exists(ui_path):
        print(f"‚ùå ERROR: No encuentro {ui_path}")
        print("üí° Aseg√∫rate de que:")
        print("   1. Existe la carpeta 'ui'")
        print("   2. El archivo 'login.ui' est√° dentro")
        print("   3. Est√° en la misma carpeta que este script")
        input("Presiona Enter para salir...")
        return
    
    if not os.path.exists(fondo_path):
        print(f"‚ö†Ô∏è Advertencia: No encuentro {fondo_path}")
    
    try:
        window = LoginWindow()
        window.show()
        sys.exit(app.exec_())
    except Exception as e:
        print(f"‚ùå Error cr√≠tico: {e}")
        input("Presiona Enter para salir...")

if __name__ == "__main__":
    main()